### ISSUE TITLE
`pulumi.all` won't work with `Output<string>` from `@pulumi/random`

### ISSUE THREAD

**Issue by oleksandr-kinship** [pulumi/pulumi#13060]
### What happened?

Tried to combine two values for user creation with `pulumi.all`, find out that it's work incorrect
In my example I have 
type user, then define Array<user>, and try using `map` create for these users appropriate resources, during debug find out that `map` works correctly iterate over all Array, but in loop using `pulumi.all` if `password` was generated by `random` lib it's not going to `pulumi.all`, step just skipped

### Expected Behavior

pulumi.all will merge outputs in single object and I will able to use few values for resource creation

### Steps to reproduce

this example code thal also won't work, because name string(can be Output<string>, it's fine) but password Output<string> from `random` lib

```
const name = "John";
const password = new random.RandomPassword("password", { length: 10 }).result;

const combinedValues = pulumi.all([name, password]).apply(([nameValue, passwordValue]) => {
  // Use the combined values here
  console.log(nameValue, passwordValue);

  // You can perform any further operations on the combined values
  const combinedString = `${nameValue}-${passwordValue}`;
  console.log(combinedString);

  // Return the combined value or perform additional actions
  return combinedString;
});  
```

in case I change `password` variable to Output<string> but not from another method, e.g using 
`interpolate`testPassword``
it will work as expected

### Output of `pulumi about`

```
CLI          
Version      3.45.0
Go Version   go1.19.2
Go Compiler  gc

Plugins
NAME        VERSION
aws         4.38.1
docker      3.6.1
kubernetes  3.27.1
newrelic    5.11.0
nodejs      unknown
random      4.13.2

Host     
OS       darwin
Version  13.2
Arch     arm64

This project is written in nodejs: executable='/opt/homebrew/bin/node' version='v19.4.0'

Backend        
Name           C11964
URL            <state>
User           olerudenko
Organizations  

Dependencies:
NAME                                              VERSION
@pulumi/aws                                       4.38.1
@pulumi/docker                                    3.6.1
@pulumi/kubernetes                                3.27.1
@pulumi/newrelic                                  5.11.0
@pulumi/pulumi                                    3.68.0
@pulumi/random                                    4.13.2
ts-node                                           7.0.1
typescript                                        4.9.5
@types/node                                       14.18.47
@typescript-eslint/eslint-plugin                  5.59.6
@typescript-eslint/parser                         5.59.6
eslint                                            8.40.0
eslint-config-prettier                            8.8.0
eslint-plugin-prettier                            4.2.1
prettier                                          2.8.8
```

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by RobbieMcKinstry** [comment]
Hi @oleksandr-kinship , just to confirm, are you experiencing this issue when running `pulumi up` or `pulumi preview`? 
I would expect the `apply` to run only during `up`, because during `preview`, there is no value for `passwordValue`, since it will not be created until you run `up`.

**Comment by RobbieMcKinstry** [comment]
I was unable to reproduce the bug using the program you provided on a newer version of Pulumi. I see you're using v3.45.0. I used 3.67.0, while the latest version is 3.69.0.

I ran the program using `preview` and again with `up`, and it worked as expected. During `preview`, the `apply` function was not executed, and there are no console logs, because the `pulumi.all()` function never resolves, since `password` is not known at the time of the `apply`.

However, when I ran `up`, I did get two log lines from inside of the `apply`. Since `password` will be created on line 5, `pulumi.all()` resolves, which allows the `apply` function to run.

* [Here's my terminal output from the `pulumi preview execution.](https://app.warp.dev/block/7bxiad1wqGvpTqS5lYo7KV)
* [Here's my terminal output from the `pulumi up` execution.](https://app.warp.dev/block/wqnztblfxE7RqeVIVEWpxR)
